### ISSUE TITLE
LoadBalancer for external databases?

### ISSUE THREAD

**Issue by kvaps** [wireapp/wire-server-deploy#196]
Hi, I've deployed wire-server with [external databases](https://docs.wire.com/how-to/install/monitoring.html) using official installation guide, then I created endpoints for these external services:

```yaml
---
kind: "Service"
apiVersion: "v1"
metadata:
  name: "cassandra-external"
spec:
  clusterIP: None
  ports:
    - name: "cassandra"
      protocol: "TCP"
      port: 9042
      targetPort: 9042
---
kind: "Endpoints"
apiVersion: "v1"
metadata:
  name: "cassandra-external"
subsets:
  - addresses:
      - ip: "10.9.8.11"
    ports:
      - port: 9042
        name: "cassandra"
  - addresses:
      - ip: "10.9.8.12"
    ports:
      - port: 9042
        name: "cassandra"
  - addresses:
      - ip: "10.9.8.13"
    ports:
      - port: 9042
        name: "cassandra"
---
kind: "Service"
apiVersion: "v1"
metadata:
  name: "elasticsearch-external"
spec:
  clusterIP: None
  ports:
    - name: "elasticsearch"
      protocol: "TCP"
      port: 9200
      targetPort: 9200
---
kind: "Endpoints"
apiVersion: "v1"
metadata:
  name: "elasticsearch-external"
subsets:
  - addresses:
      - ip: "10.9.8.21"
    ports:
      - port: 9200
        name: "cassandra"
  - addresses:
      - ip: "10.9.8.22"
    ports:
      - port: 9200
        name: "cassandra"
  - addresses:
      - ip: "10.9.8.23"
    ports:
      - port: 9200
        name: "cassandra"
---
kind: "Service"
apiVersion: "v1"
metadata:
  name: "minio-external"
spec:
  clusterIP: None
  ports:
    - name: "minio"
      protocol: "TCP"
      port: 9000
      targetPort: 9000
---
kind: "Endpoints"
apiVersion: "v1"
metadata:
  name: "minio-external"
subsets:
  - addresses:
      - ip: "10.9.8.31"
    ports:
      - port: 9000
        name: "minio"
  - addresses:
      - ip: "10.9.8.32"
    ports:
      - port: 9000
        name: "minio"
  - addresses:
      - ip: "10.9.8.33"
    ports:
      - port: 9000
        name: "minio"
---
kind: "Service"
apiVersion: "v1"
metadata:
  name: "redis-external"
spec:
  clusterIP: None
  ports:
    - name: "elasticsearch"
      protocol: "TCP"
      port: 6379
      targetPort: 6379
---
kind: "Endpoints"
apiVersion: "v1"
metadata:
  name: "redis-external"
subsets:
  - addresses:
      - ip: "10.9.8.41"
    ports:
      - port: 6379
        name: "redis"
  - addresses:
      - ip: "10.9.8.42"
    portsж
      - port: 6379
        name: "redis"
  - addresses:
      - ip: "10.9.8.43"
    ports:
      - port: 6379
        name: "redis"
```

After, I found out that official helm charts also using the same logic:
- [cassandra-external](https://github.com/wireapp/wire-server-deploy/blob/develop/charts/cassandra-external/templates/endpoint.yaml)
- [elasticsearch-external](https://github.com/wireapp/wire-server-deploy/blob/develop/charts/elasticsearch-external/templates/endpoint.yaml)
- [minio-external](https://github.com/wireapp/wire-server-deploy/blob/develop/charts/minio-external/templates/endpoint.yaml)

Problem is that created manually endpoints have no any healthcheck mechanism and they are not high available by the default, thus there should be some logic implemented on the client side or another load balancer. 

My question does wire-server this logic implemented, or should I to use some external loadbalancer eg. harpoxy to route the traffic to these services?

**Comment by jschaul** [comment]
For cassandra, load balancer logic is not needed, it would actually be harmful. Cassandra driver applies a client-side load balancing. External Load Balancer may become a Single Point of Failure. When pods using cassandra start (i.e. `brig`, `gundeck`, `galley` and `spar`), they contact cassandra based on the IPs given. After the first contact, any changes to cassandra topology is transmitted to the cassandra driver on the client (inside e.g. `brig`). So for cassandra, it is actually highly-available as long as at least one IP address inside the `cassandra-external` chart is reachable at all times (since a pod may restart or new pods may come up during an upgrade). The cassandra driver also establishes a connection pool directly to multiple cassandra nodes and keeps that alive and updated.

For elasticsearch and minio, there is no smart client-side driver logic like there is for cassandra. All the code does is send http requests to a DNS name, and retry them a few times if they fail. What exact failure scenario are you considering, where you think a load balancer would help?

**Comment by kvaps** [comment]
Thanks for the comprehensive answer.

I deployed wire-server using the next scheme:
![](https://docs.wire.com/_images/architecture-server-ha.png)

I was considering case when one physical node will go down it shouldn't affect on wire-server normal work.

From your answer and according fact that Kubernetes uses roundrobin, I realized that it shouldn't be problem.

> All the code does is send http requests to a DNS name, and retry them a few times if they fail

Does this statement also correct for wire clients? Eg. webapp or mobile application could try downloading some files from S3, but minio located behind nginx-ingress-controller which have no healthchecks for endpoints.
